test/e2e/network/netpol/network_policy.go

```diff
@@ -0,0 +1,904 @@
+and what is happening in practice:
+
+		Oct 19 10:34:16.907: INFO: expected:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	X	.	.	.	.	.	.	.	.
+		x/b	X	.	.	.	.	.	.	.	.
+		x/c	X	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	X	.	.	.	.	.	.	.	.
+		z/b	X	.	.	.	.	.	.	.	.
+		z/c	X	.	.	.	.	.	.	.	.
+
+		Oct 19 10:34:16.907: INFO: observed:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	X	.	.	.	.	.	.	.	.
+		x/b	X	.	.	.	.	.	.	.	.
+		x/c	X	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	X	.	.	.	.	.	.	.	.
+		z/b	X	.	.	.	.	.	.	.	.
+		z/c	X	.	.	.	.	.	.	.	.
+
+		Oct 19 10:34:16.907: INFO: comparison:
+
+		-	x/a	x/b	x/c	y/a	y/b	y/c	z/a	z/b	z/c
+		x/a	.	.	.	.	.	.	.	.	.
+		x/b	.	.	.	.	.	.	.	.	.
+		x/c	.	.	.	.	.	.	.	.	.
+		y/a	.	.	.	.	.	.	.	.	.
+		y/b	.	.	.	.	.	.	.	.	.
+		y/c	.	.	.	.	.	.	.	.	.
+		z/a	.	.	.	.	.	.	.	.	.
+		z/b	.	.	.	.	.	.	.	.	.
+		z/c	.	.	.	.	.	.	.	.	.
+*/
+var _ = network.SIGDescribe("Netpol [LinuxOnly]", func() {
+	f := framework.NewDefaultFramework("netpol")
+
+	ginkgo.Context("NetworkPolicy between server and client", func() {
+		ginkgo.BeforeEach(func() {
+			if useFixedNamespaces {
+				_ = initializeResources(f)
+
+				_, _, _, model, k8s := getK8SModel(f)
+				framework.ExpectNoError(k8s.CleanNetworkPolicies(model.NamespaceNames), "unable to clean network policies")
+				err := wait.Poll(waitInterval, waitTimeout, func() (done bool, err error) {
+					for _, ns := range model.NamespaceNames {
+						netpols, err := k8s.ClientSet.NetworkingV1().NetworkPolicies(ns).List(context.TODO(), metav1.ListOptions{})
+						framework.ExpectNoError(err, "get network policies from ns %s", ns)
+						if len(netpols.Items) > 0 {
+							return false, nil
+						}
+					}
+					return true, nil
+				})
+				framework.ExpectNoError(err, "unable to wait for network policy deletion")
+			} else {
+				framework.Logf("Using %v as the default dns domain for this cluster... ", framework.TestContext.ClusterDNSDomain)
+				framework.ExpectNoError(initializeResources(f), "unable to initialize resources")
+			}
+		})
+
+		ginkgo.AfterEach(func() {
+			if !useFixedNamespaces {
+				_, _, _, model, k8s := getK8SModel(f)
+				framework.ExpectNoError(k8s.deleteNamespaces(model.NamespaceNames), "unable to clean up netpol namespaces")
+			}
+		})
+
+		ginkgo.It("should support a 'default-deny-ingress' policy [Feature:NetworkPolicy]", func() {
+			nsX, _, _, model, k8s := getK8SModel(f)
+			policy := GetDenyIngress("deny-ingress")
+			CreatePolicy(k8s, policy, nsX)
+
+			reachability := NewReachability(model.AllPods(), true)
+			reachability.ExpectPeer(&Peer{}, &Peer{Namespace: nsX}, false)
+
+			ValidateOrFail(k8s, model, &TestCase{FromPort: 81, ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachability})
+		})
+
+		ginkgo.It("should support a 'default-deny-all' policy [Feature:NetworkPolicy]", func() {
```

Comment:
@thockin @bowei I brought this up in sig today, this is I think the ideal way to define all these tests in the future , i.e. as json/yaml raw ...   we did it for this one case as a start but can expand to the others in a potential follow on later if folks decide its more readable

Reply:
Thoughts:

Main downside is that a mistake in JSON will not compile-fail and may not even visibly runtime fail.  That's pretty significant

I might suggest a different take - there's a lot of boilerplate here.  Can you condense all of this down to *JUST* the bits that are different between each testcase.  Maybe even shoot for tabular form.

E.g., can this be reduced to:

```
{
    it: "should support a 'default-deny-all' policy [Feature:NetworkPolicy]",
    name: "deny-all-tcp-allow-dns",
    podSelector: select(),
    egress: egress(ports(port(53, "UDP")), policyTypes("Ingress", "Egress"), Peers()),
    expectReachable: ...
    expectUnreachable: ...
}, {
   // more
```

You can use all sorts of clever variadic functions and funtions that return `func(*NetworkPolicySpec)` to make a "DSL" of sorts that retains type-safety and brevity and makes it super-easy to see what is important in this testcase.  As you loop over the table you re-assemble the policy you want.

The reductivist test

Reply:
We originally had a full blown dsl, but roughly qouting @caseydavenport  ...

```
there are already 2 ways to express network policies: structs and json.  adding a third way 
creates cognitive load without being entiirely necessary 
```

so we just stuck w/ go structs for this PR :).  but i think maybe we could convince folks that A DSL would be good long term.  Was thinking to do that in a follow on ? @mattfenwick has i think some ideas there.

Reply:
I did push back against the original suggestion to invent a new way to express policies, but I am happy to be proven wrong!

It if were me, I'd try to avoid the extra layer of abstraction that a new network-policy-builder "language" would introduce. Makes it harder to read / write new tests if you need to learn or modify some new NP declaration framework rather than using the raw API structs themselves, with the added benefit of it being more obvious exactly what the NP under test is.

Maybe some sort of compromise? 

```
// Returns a NP with the name / namespace, the rest unset.
func newBasicPolicy(name, namespace string) NetworkPolicy {}

```

then

```
npUnderTest = newBasicPolicy(name, ns)
npUnderTest.Spec.Selector = . . .
npUnderTest.Spec.Ingress = . . .
npUnderTest.Spec.Egress = . . .
```

At least then you are dealing directly with the structs, but hiding most of the boilerplate within a helper function?

Reply:
@jayunit100 it sounds like I might be outnumbered, so feel free to bypass my objections :smile:

Reply:
Ok... DSL is definetly cool but.. ok to move this to a follow on issue? 

Seems like it's an iteration on top of this,  since the KEP which we're implementing specifically concluded "no DSL ... yet".

Reply:
btw casey what were doing in this PR is exactly what you suggest above :), after all it was your idea to use raw structs.

```
                       policy := GetAllowIngressByPod("allow-client-a-via-pod-selector", map[string]string{"pod": "a"}, allowedPodLabels)
			policy.Spec.Egress = ... 
			policy.Spec.PolicyTypes = []networkingv1.PolicyType{networkingv1.PolicyTypeEgress, networkingv1.PolicyTypeIngress}
			nsX, _, _, model, k8s := getK8SModel(f)
			CreatePolicy(k8s, policy, nsX)

			reachabilityPort80 := NewReachability(model.AllPods(), true)
			reachabilityPort80.ExpectAllIngress(NewPodString(nsX, "a"), false)
			reachabilityPort80.Expect(NewPodString(nsX, "b"), NewPodString(nsX, "a"), true)
			ValidateOrFail(k8s, model, &TestCase{FromPort: 81, ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort80})
``` 

:)